### PR TITLE
Add ARM support for WebJobs api.

### PR DESCRIPTION
**Pull request by amitapl**
Also add settings property to WebJob.

### CODE REVIEW

File: Kudu.Services/Arm/ArmActionSelector.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Kudu.Services.Arm
+{
+    public class ArmActionSelector : ApiControllerActionSelector
+    {
+        public override HttpActionDescriptor SelectAction(HttpControllerContext controllerContext)
+        {
+            if (controllerContext.Request.Method == HttpMethod.Put && ArmUtils.IsArmRequest(controllerContext.Request))
+            {
+                controllerContext.RouteData.Values["action"] += "Arm";
```
Comment (davidebbo) [inline Kudu.Services/Arm/ArmActionSelector.cs]: Comment would be good here is this looks pretty cryptic without context.

File: Kudu.Services/Arm/ArmControllerConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using System.Web.Http.Controllers;
+
+namespace Kudu.Services.Arm
+{
+    [AttributeUsage(AttributeTargets.Class)]
+    public sealed class ArmControllerConfiguration : Attribute, IControllerConfiguration
+    {
+        public void Initialize(HttpControllerSettings controllerSettings, HttpControllerDescriptor controllerDescriptor)
+        {
+            controllerSettings.Services.Replace(typeof(IHttpActionSelector), new ArmActionSelector());
```
Comment (davidebbo) [inline Kudu.Services/Arm/ArmControllerConfiguration.cs]: Should that use a singleton instance?

Comment (davidebbo) [inline Kudu.Services/Arm/ArmControllerConfiguration.cs]: Well, object looks cheap so it may not matter.

File: Kudu.Services/Arm/ArmUtils.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+                armEntry.Name = String.Empty;
+                for (int i = 8; i < idTokens.Length; i += 2)
+                {
+                    armEntry.Name += "/" + idTokens[i];
+                }
+            }
+
+            IEnumerable<string> values;
+            if (request.Headers.TryGetValues(GeoLocationHeaderKey, out values))
```
Comment (suwatch) [inline Kudu.Services/Arm/ArmUtils.cs]: it gets this far, it must have GeoLocationHeader.   I would skip TryGetValue.

File: Kudu.Services/Arm/ArmUtils.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        private static ArmEntry<T> Create<T>(T o, HttpRequestMessage request, bool isChild = false) where T : INamedObject
+        {
+            var armEntry = new ArmEntry<T>()
+            {
+                Properties = o
+            };
+
+            // In Azure ARM requests, the referrer is the current id
+            Uri referrer = request.Headers.Referrer;
+            if (referrer != null)
```
Comment (suwatch) [inline Kudu.Services/Arm/ArmUtils.cs]: This is one of the contract that it came from CSM.   Maybe we should add this check in IsArmRequest - and no need to check here.

File: Kudu.Services/Arm/ArmActionSelector.cs (hunk 1 above)
Comment (suwatch) [inline Kudu.Services/Arm/ArmActionSelector.cs]: Especially, where it is being consumed.  I somehow can't find it.

File: Kudu.Services/Arm/ArmControllerConfiguration.cs (hunk 2 above)
Comment (davidebbo) [inline Kudu.Services/Arm/ArmControllerConfiguration.cs]: Yes, it's tricky. It modifies the name of the action method that we're binding to. So foo --> FooArm. That's why you don't find "arm" literal anywhere else. Definitely calls for comment!

File: Kudu.Services/Arm/ArmUtils.cs (hunk 3 above)
Comment (amitapl) [inline Kudu.Services/Arm/ArmUtils.cs]: I would really rather be safe here

File: Kudu.Services/Arm/ArmUtils.cs (hunk 4 above)
Comment (amitapl) [inline Kudu.Services/Arm/ArmUtils.cs]: Again I really rather be safe

**Comment by amitapl** [comment]
Added some comments

**Comment by amitapl** [comment]
shipit?

**Comment by davidebbo** [comment]
Yep.

**Pull request closed without merging**